angr/engines/pcode/engine.py

```diff
@@ -0,0 +1,200 @@
+                    insn_bytes=insn_bytes,
+                    thumb=thumb,
+                    size=size,
+                    num_inst=num_inst,
+                    extra_stop_points=extra_stop_points,
+                )
+
+            if irsb.size == 0:
+                if (
+                    irsb.jumpkind == "Ijk_NoDecode"
+                    and not self.state.project.is_hooked(irsb.addr)
+                ):
+                    raise errors.SimIRSBNoDecodeError(
+                        "IR decoding error at %#x. You can hook this instruction with "
+                        "a python replacement using project.hook"
+                        "(%#x, your_function, length=length_of_instruction)."
+                        % (addr, addr)
+                    )
+
+                raise errors.SimIRSBError("Empty IRSB passed to HeavyVEXMixin.")
+
+            # check permissions, are we allowed to execute here? Do we care?
+            if o.STRICT_PAGE_ACCESS in self.state.options:
+                try:
+                    perms = self.state.memory.permissions(addr)
+                except errors.SimMemoryError:
+                    raise errors.SimSegfaultError(addr, "exec-miss")
+                else:
+                    if not perms.symbolic:
+                        perms = self.state.solver.eval(perms)
+                        if not perms & 4 and o.ENABLE_NX in self.state.options:
+                            raise errors.SimSegfaultError(addr, "non-executable")
+
+            # FIXME
+            # self.state.scratch.set_tyenv(irsb.tyenv)
+
+            self.state.scratch.irsb = irsb
+
+            # fill in artifacts
+            successors.artifacts["irsb"] = irsb
+            successors.artifacts["irsb_size"] = irsb.size
+            successors.artifacts["irsb_direct_next"] = irsb.direct_next
+            successors.artifacts["irsb_default_jumpkind"] = irsb.jumpkind
+            successors.artifacts["insn_addrs"] = []
+
+            try:
+                self.handle_pcode_block(irsb)
+            except errors.SimReliftException as e:
+                self.state = e.state
+                if insn_bytes is not None:
+                    raise errors.SimEngineError(
+                        "You cannot pass self-modifying code as insn_bytes!!!"
+                    )
+                new_ip = self.state.scratch.ins_addr
+                if size is not None:
+                    size -= new_ip - addr
+                if num_inst is not None:
+                    num_inst -= self.state.scratch.num_insns
+                addr = new_ip
+
+                # clear the stage before creating the new IRSB
+                self.state.scratch.dirty_addrs.clear()
+                irsb = None
+
+            except errors.SimError as ex:
+                ex.record_state(self.state)
+                raise
+            # FIXME:
+            # except VEXEarlyExit:
+            #     break
+            else:
+                break
```

Comment:
I'm not a big fan of this C-style `while True: <many lines of code>... break` paradigm. Can you rewrite this method so that it's easier to read and understand? You may want to add a new method instead.

Reply:
This is how the VEX engine has implemented this (`HeavyVEXMixin.process_successors`), so I followed it for consistency. I can find a way to make this easier to read, however.

Reply:
I've done initial refactoring of this method, while preserving the logic. There is still room for improvement, but I think it is more readable. It looks like this now:

```python
        # Lift and process the block to get successors, retry if necessary
        finished = False
        while not finished:
            self._lift_irsb()
            self._probe_access()
            self._store_successor_artifacts(successors)
            finished = self._process_irsb()

        self._process_successor_exits(successors)
        successors.processed = True
```